shell/platform/windows/external_texture_gl.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/windows/external_texture_gl.h"
+
+#include <EGL/egl.h>
+#include <GLES2/gl2.h>
+#include <GLES2/gl2ext.h>
+
+#include <iostream>
+
+namespace {
+
+typedef void (*glGenTexturesProc)(GLsizei n, GLuint* textures);
+typedef void (*glDeleteTexturesProc)(GLsizei n, const GLuint* textures);
+typedef void (*glBindTextureProc)(GLenum target, GLuint texture);
+typedef void (*glTexParameteriProc)(GLenum target, GLenum pname, GLint param);
+typedef void (*glTexImage2DProc)(GLenum target,
+                                 GLint level,
+                                 GLint internalformat,
+                                 GLsizei width,
+                                 GLsizei height,
+                                 GLint border,
+                                 GLenum format,
+                                 GLenum type,
+                                 const void* data);
+
+// A struct containing pointers to resolved gl* functions.
+struct GlProcs {
+  glGenTexturesProc glGenTextures;
+  glDeleteTexturesProc glDeleteTextures;
+  glBindTextureProc glBindTexture;
+  glTexParameteriProc glTexParameteri;
+  glTexImage2DProc glTexImage2D;
+  bool valid;
+};
+
+static const GlProcs& GlProcs() {
+  auto glGenTextures =
+      reinterpret_cast<glGenTexturesProc>(eglGetProcAddress("glGenTextures"));
+  auto glDeleteTextures = reinterpret_cast<glDeleteTexturesProc>(
+      eglGetProcAddress("glDeleteTextures"));
+  auto glBindTexture =
+      reinterpret_cast<glBindTextureProc>(eglGetProcAddress("glBindTexture"));
+  auto glTexParameteri = reinterpret_cast<glTexParameteriProc>(
+      eglGetProcAddress("glTexParameteri"));
+  auto glTexImage2D =
+      reinterpret_cast<glTexImage2DProc>(eglGetProcAddress("glTexImage2D"));
```

Comment:
You're doing all of these lookups every time the function is called, which defeats the point of having the struct below be `static`.

The basic structure of this function should be something like:
```
static bool sInitialized = false;
static struct GlProcs sProcs = {};
if (!sInitialized) {
  [ do all the lookup to populate sProcs ]
  sInitialized = true;
}
return sProcs;
```

Reply:
I agree. But a test showed that the compiler appears to be smart enough to not call `eglGetProcAddress` again " (as it's const here).
But it might to easier to just change it as opposed to finding evidence ;)

Reply:
I'm curious how you did that test, because I don't see how the compiler could be "smart enough" to optimize away calls to a function that could have arbitrary side effects...

Reply:
I created a "spy" function logging all invocations to `eglGetProcAddress`.

Reply:
I'm not sure what that means exactly. (In fact, the possibility of the function doing logging is an illustration of exactly why I would not expect the optimization you are describing to be possible.)